engine/src/bloom.rs

```diff
@@ -0,0 +1,91 @@
+//! Based on Parity Common Eth Bloom implementation
+//! Link: <https://github.com/paritytech/parity-common/blob/master/ethbloom/src/lib.rs>
+//!
+//! Reimplemented here since there is a large miss match in types and dependencies.
+#![allow(clippy::expl_impl_clone_on_copy)]
+
+use aurora_engine_types::parameters::engine::ResultLog;
+use fixed_hash::construct_fixed_hash;
+use impl_serde::impl_fixed_hash_serde;
+
+use aurora_engine_sdk::keccak;
+
+#[cfg(not(feature = "borsh-compat"))]
+use borsh::{BorshDeserialize, BorshSerialize};
+#[cfg(feature = "borsh-compat")]
+use borsh_compat::{self as borsh, BorshDeserialize, BorshSerialize};
+
+const BLOOM_SIZE: usize = 256;
+const BLOOM_BITS: u32 = 3;
+
+construct_fixed_hash! {
+    /// Bloom hash type with 256 bytes (2048 bits) size.
+    #[derive(BorshSerialize, BorshDeserialize)]
+    pub struct Bloom(BLOOM_SIZE);
+}
+
+impl_fixed_hash_serde!(Bloom, BLOOM_SIZE);
+
+/// Returns log2.
+const fn log2(x: usize) -> u32 {
+    if x <= 1 {
+        return 0;
+    }
+
+    let n = x.leading_zeros();
+    usize::BITS - n
+}
+
+impl Bloom {
+    /// Add a new element to the bloom filter
+    #[allow(clippy::as_conversions)]
+    pub fn accrue(&mut self, input: &[u8]) {
+        let m = self.0.len();
+        let bloom_bits = m * 8;
+        let mask = bloom_bits - 1;
+        let bloom_bytes = (log2(bloom_bits) + 7) / 8;
+        let hash = keccak(input);
+        let mut ptr = 0;
+
+        for i in 0..BLOOM_BITS {
+            let _ = i;
```

Comment:
Why? Could you please explain this trick?

Reply:
Sorry for not clarifying this before. This file is coming from the Refiner, [here](https://github.com/aurora-is-near/borealis-engine-lib/blob/main/refiner-types/src/bloom.rs). I plan to remove it from Refiner later and consume it from the engine.